Solving the issues in small PR #17

plantton opened the pull request:
- Remove redundant `return` functions  
- Add an `IF` statement to function `.attach_migration`: Only working for SQLite connections now.   
- Preserve all data/data tables, while merging two objects.  
- Add a reset function to revert row selection.

lgatto commented on NEWS.md:
```diff
@@ -0,0 +1,4 @@
+# MsBackendSql 0.1.0
+
+Changes in 0.5.0  
```
The version here is wrong. If you want to change the version number, it should be the last one, and this needs also be done in the `DESCRIPTION` file.

plantton replied:
Ok. I will check it.

plantton replied:
> The version here is wrong. If you want to change the version number, it should be the last one, and this needs also be done in the `DESCRIPTION` file.

I changed the current version to `0.5.0` in commit [02c29cc](https://github.com/rformassspectrometry/MsBackendSql/pull/21/commits/02c29cc5ea47c423dea4d46310a081be774f89e0).

lgatto commented on R/MsBackendSqlDb-functions.R:
```diff
@@ -20,11 +20,11 @@ MsBackendSqlDb <- function(dbcon) {
         slot(x, "dbcon", check = FALSE) <- dbConnect(SQLite(), 
                                                tempfile(fileext = ".db"))
-        return(x)
+        x
```
I prefer to keep `return()` here, as it makes clear that the function can return in the middle of its body.

plantton replied:
> I prefer to keep `return()` here, as it makes clear that the function can return in the middle of its body.

modified in commit [277facd](https://github.com/rformassspectrometry/MsBackendSql/pull/21/commits/277facd25cdc215966146f54dcc1b344f8562eb7).

lgatto commented on R/MsBackendSqlDb-functions.R:
```diff
@@ -20,11 +20,11 @@ MsBackendSqlDb <- function(dbcon) {
       x <- new("MsBackendSqlDb")
       slot(x, "dbcon", check = FALSE) <- dbcon
-      return(x)
+      x
```
Same as above.

plantton replied:
> Same as above.

modified in commit [277facd](https://github.com/rformassspectrometry/MsBackendSql/pull/21/commits/277facd25cdc215966146f54dcc1b344f8562eb7).

lgatto commented on R/MsBackendSqlDb-functions.R:
```diff
@@ -303,13 +303,13 @@ MsBackendSqlDb <- function(dbcon) {
             objects[[1]] <- .attach_migration(objects[[1]], objects[[i]])
         }
-        return(objects[[1]])
+        objects[[1]]
```
Same as above.

plantton replied:
> Same as above.

modified in commit [277facd](https://github.com/rformassspectrometry/MsBackendSql/pull/21/commits/277facd25cdc215966146f54dcc1b344f8562eb7).

lgatto commented on R/MsBackendSqlDb-functions.R:
```diff
@@ -303,13 +303,13 @@ MsBackendSqlDb <- function(dbcon) {
             res <- .attach_migration(res, objects[[i]])
         }
-        return(res)
+        res
```
Same as above.

plantton replied:
> Same as above.

modified in commit [277facd](https://github.com/rformassspectrometry/MsBackendSql/pull/21/commits/277facd25cdc215966146f54dcc1b344f8562eb7).

lgatto commented on R/MsBackendSqlDb-functions.R:
```diff
@@ -435,7 +441,7 @@ MsBackendSqlDb <- function(dbcon) {
                                                    res@dbtable, 
                                                    " where _pkey = ?"))
-        return(res)
+        res
```
Here no `return()` is fine.

plantton replied:
> Here no `return()` is fine.

modified in commit [277facd](https://github.com/rformassspectrometry/MsBackendSql/pull/21/commits/277facd25cdc215966146f54dcc1b344f8562eb7).

lgatto commented on R/MsBackendSqlDb.R:
```diff
@@ -557,7 +557,7 @@ setMethod("split", "MsBackendSqlDb", function(x, f, drop = FALSE, ...) {
         output[[i]] <- x
     }
-    return(output)
+    output
```
Here no `return()` is fine.

lgatto commented on R/MsBackendSqlDb-functions.R:
```diff
@@ -357,7 +360,7 @@ MsBackendSqlDb <- function(dbcon) {
         ## The writing operation increases "1" for the merged instance.
         x@modCount <- max(x@modCount, y@modCount) + 1L
-        return(x) } else {
+        x} else {
```
Please leave `return()` here.

lgatto replied:
Add a new line before `} else ...`

plantton replied:
modified in commit [277facd](https://github.com/rformassspectrometry/MsBackendSql/pull/21/commits/277facd25cdc215966146f54dcc1b344f8562eb7).

lgatto commented on R/MsBackendSqlDb-functions.R:
```diff
@@ -327,13 +327,16 @@ MsBackendSqlDb <- function(dbcon) {
         identical(x@dbtable, y@dbtable)) {
         x@rows <- c(x@rows, y@rows)
-        return(x) } else if (identical(x@dbcon@dbname, y@dbcon@dbname) && 
+        x } else if (identical(x@dbcon@dbname, y@dbcon@dbname) && 
```
Please leave `return()` here.

lgatto replied:
Add a new line before `} else ...`

plantton replied:
modified in commit [277facd](https://github.com/rformassspectrometry/MsBackendSql/pull/21/commits/277facd25cdc215966146f54dcc1b344f8562eb7).

lgatto commented:
The commit for the `reset()` method looks good to me. Let's wait for a review by @jorainer

jorainer commented on R/MsBackendSqlDb-functions.R:
```diff
@@ -550,4 +554,18 @@ MsBackendSqlDb <- function(dbcon) {
+    ## We fetch the number of rows for SQLite table
+    rowNum <- dbGetQuery(x@dbcon, paste0("SELECT COUNT(*) FROM ", x@dbtable))
+    ## Then link `rows` index to SQLite table
+    slot(x, "rows", check = FALSE) <- 1L:rowNum[1, 1]
```
Please try to use `seq_along`, `seq_len` etc instead of  `1:` wherever possible (what if your table is empty? you would get `1:0` although `rows` should then be `integer(0)`). Please use `seq_len(rowNum[1, 1])` instead.

plantton replied:
Thank you. That's a very good point!

jorainer commented on R/MsBackendSqlDb-functions.R:
```diff
@@ -550,4 +554,18 @@ MsBackendSqlDb <- function(dbcon) {
+#' 
+#' @noRd
+.reset_row_indices <- function(x) {
+    ## We fetch the number of rows for SQLite table
```
Also, comments are nice, but here they are just too much because it is easily possible to understand what you are doing from the code itself. Try to use only comments in the code if absolutely necessary.

lgatto replied:
Comments shouldn't explain what the code does. The code should be self-explanatory; if it isn't, it's a sign that it should be improved. Comments should explain why things work in a certain way, focus on the higher level logic, explain some side effects that aren't obvious from the code itself, ...

jorainer reviewed (changes requested):
Thanks for the PR @plantton ! Looks great. Only minor comments. And please update/fix the documentation of the `MsBackendSQL` object. It looks like it's still the very first version I wrote. Please update:
- describe what the `reset` method does. This is something that other backends can't do (i.e. undo row filtering).
- provide some background information, such as where the data is stored, what data is kept in memory etc.
- add yourself as author

lgatto commented:
@plantton - have @jorainer's points been addressed?

jorainer reviewed (commented):
I think you addressed all points regarding the code and the functionality @plantton , but you still need to update/extend the documentation of your backend (*MsBackendSqlDb.R* file lines 4-91. Please check if the text is still valid and correct/expand accordingly.

plantton commented:
> I think you addressed all points regarding the code and the functionality @plantton , but you still need to update/extend the documentation of your backend (_MsBackendSqlDb.R_ file lines 4-91. Please check if the text is still valid and correct/expand accordingly.

I will try to improve the documents in the following days. And I also think it has to be adjusted to the latest version of `Spectra`.

jorainer commented:
Great. When you're done please re-request me to review - so I know I should check it again.

lgatto commented on NEWS.md:
```diff
@@ -1,6 +1,10 @@
-	- rename asDataFrame to spectraData.  
+	- rename asDataFrame to spectraData. 
+	
+## Changes in 0.5.1  
```
The 0.5.1 section much come above 0.5.0

lgatto commented:
The vignette fails to build with
```
Quitting from lines 61-65 (MsBackendSqlDb.Rmd) 
Error: Error: processing vignette 'MsBackendSqlDb.Rmd' failed with diagnostics:
unused argument (dbcon = conn)
--- failed re-building ‘MsBackendSqlDb.Rmd’
```

plantton commented:
> The vignette fails to build with
> 
> ```
> Quitting from lines 61-65 (MsBackendSqlDb.Rmd) 
> Error: Error: processing vignette 'MsBackendSqlDb.Rmd' failed with diagnostics:
> unused argument (dbcon = conn)
> --- failed re-building ‘MsBackendSqlDb.Rmd’
> ```

Yes. I'm still reviewing all the functions.
